### ISSUE TITLE
Logging validation loss without library code hacks

### ISSUE THREAD

**Issue by mfernezir** [open-mmlab/mmsegmentation#1396]
I'm using the approach with two workflows, which is suggested as an option to log the validation loss.

Namely, in the notebook tutorial:

```
from mmseg.datasets import build_dataset
from mmseg.models import build_segmentor
from mmseg.apis import train_segmentor
import copy

cfg.workflow = [('train', 1), ('val', 1)]
datasets = [build_dataset(cfg.data.train)]

if len(cfg.workflow) == 2:
  val_dataset = copy.deepcopy(cfg.data.val)
  val_dataset.pipeline = cfg.data.train.pipeline
  datasets.append(build_dataset(val_dataset))

model = build_segmentor(
    cfg.model)

model.CLASSES = datasets[0].CLASSES

mmcv.mkdir_or_exist(osp.abspath(cfg.work_dir))
train_segmentor(model, datasets, cfg, distributed=False, validate=True, 
                meta=dict())
```

However, this doesn't actually log both the training and validation loss, as it is supposed to. There are two closed issues, #310 and #483, that additionally modify https://github.com/open-mmlab/mmsegmentation/blob/master/mmseg/models/segmentors/base.py#L148 to achieve the desired effect, as described here: https://github.com/open-mmlab/mmsegmentation/issues/310#issuecomment-757499226

I'm currently using a small variation of this solution as a replacement in `base.py`:
```
    def val_step(self, data_batch, optimizer=None, **kwargs):
        """The iteration step during validation.

        This method shares the same signature as :func:`train_step`, but used
        during val epochs. Note that the evaluation after training epochs is
        not implemented with this method, but an evaluation hook.
        """
        losses = self(**data_batch)
        loss, log_vars = self._parse_losses(losses)
        
        for k in list(log_vars.keys()):
            log_vars["val_" + k] = log_vars[k]
            del log_vars[k]

        outputs = dict(
            loss=loss,
            log_vars=log_vars,
            num_samples=len(data_batch['img_metas']))

        return outputs
```
1. Is it possible to achieve equivalent validation loss logging without modifying the library code?
2. If not, is this solution sensible as a general approach for a pull request?

**Comment by MeowZheng** [comment]
You might implement a custom hook inherited from hook from mmcv and override `every_n_inner_iters` https://github.com/open-mmlab/mmcv/blob/e0eebebf50750139d25f1575e1a29902a672794d/mmcv/runner/hooks/hook.py#L58

**Comment by mfernezir** [comment]
@MeowZheng I'm not sure how overriding that function would help, as it only returns True or False...

Still, this issue looks like an MMSegmentation bug when running both train and val workflow modes.

Iteration runner calls the model's val_step here
https://github.com/open-mmlab/mmcv/blob/44edcdd91f8940a58c9680c609febc757a50a040/mmcv/runner/iter_based_runner.py#L77

The model's val_step returns the loss under the same key as in the training mode, which gets lost in all output logs later on.

Do you think a good solution would be to create a pull request with the above change in the val_step that adds the "val_" prefix under validation losses, or possibly with the `self._parse_function` in base.py to accept mode "val" to make it a bit cleaner? I can create a pull request if you think that makes sense, but I'm not sure if there would be some unintended consequences elsewhere. This approach works well for me to distinguish training and validation losses in the console, tensorboard, and wandb.

**Comment by MeowZheng** [comment]
I'm so sorry for the late reply and inappropriate solution. I think you are right. The validation loss was lost due to the same key for train and validation loss 

In addition, I have tried the solution which is almost the same as you proposed, like below:

```
    def val_step(self, data_batch, optimizer=None, **kwargs):
        """The iteration step during validation.

        This method shares the same signature as :func:`train_step`, but used
        during val epochs. Note that the evaluation after training epochs is
        not implemented with this method, but an evaluation hook.
        """
        losses = self(**data_batch)
        loss, log_vars = self._parse_losses(losses)
        log_vars_ = dict()
        for loss_name, loss_value in log_vars.items():
            k = loss_name + '_val'
            log_vars_[k] = loss_value

        outputs = dict(
            loss=loss,
            log_vars=log_vars_,
            num_samples=len(data_batch['img_metas']))

        return outputs
```

Please create a  pull request for this modification, and　thank you in advance.

**Comment by ZwwWayne** [comment]
Invite @HAOCHENYE to take a look

**Comment by shinpaul14** [comment]
How can we use this val_step in the most recent version??
